### PR TITLE
pr-template: Move contributor instructions to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by l0b0**
Example PR template filled in below:

## Changes

- Move contributor instructions to CONTRIBUTING.md
- Use semantically valid H2 tags for level-2 headings.

## Manual tests

Discussed changes with the [community](https://app.element.io/#/room/#nix:nixos.org).

## Sandboxed non-Linux build?

If this change is applicable to non-Linux platforms such as macOS, please indicate your [`nix.conf` `sandbox` setting](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-sandbox):

- [ ] `sandbox = true`
- [ ] `sandbox` is not set
- [ ] `sandbox = false`

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Not a fan of this, I wish we had easier ways to auto-tick them in some cases rather than removing them.

**Comment by mweinelt** [comment]
This change seems very opinionated and does not provide any motivation for why we should adopt it.

**Comment by evils** [comment]
to rehash some of the matrix chatter from yesterday:

most of the tests should get run by the CI, the contributor shouldn't need to run them locally
(but may want to, so we should point out they exist and how to use them in the contributing guidelines)

which is to say, anyone reviewing a PR will know these tests get run because CI does so
and the contributor saying they already did so adds no value to the reviewer

whether release notes have been added should be obvious from the diff
so a checkbox for that serves primarily as a reminder to write release notes
(still very valuable!)

i think the overall motivation is to
1. clarify things for new contributors
2. reduce irrelevant information presented to reviewers

which i think leaves the following points we want in the template:
1. high level summary of what the PR does
2. what tests did the contributor perform that aren't performed by the CI (manual checks, nixpkgs-review?)
3. on what platform did they they perform their actions
4. did the contributor have the contributing guidelines in mind when making these changes?
5. some non-binary indication of care having been taken in drafting the PR (mostly my opinion, but i feel like that's being lost in the current draft (contributing.md checkbox kinda asks this already though))



what about something like this?:

## Changes
- [ ] are made with CONTRIBUTING.md in mind
- [ ] include major or breaking changes

### Description
- Move contributor instructions to CONTRIBUTING.md
- Use semantically valid H2 tags for level-2 headings.

## Manual tests
- [ ] changes actually work
- [ ] ran `nixpkgs-review`

### Performed on:
- Linux
  - [ ] x86-64
  - [ ] aarch64
- Darwin
  - [ ] x86-64
  - [ ] aarch64
  - [ ] with `sandbox = true`

**Comment by mweinelt** [comment]
> most of the tests should get run by the CI, the contributor shouldn't need to run them locally
> (but may want to, so we should point out they exist and how to use them in the contributing guidelines)
> 
> which is to say, anyone reviewing a PR will know these tests get run because CI does so
> and the contributor saying they already did so adds no value to the reviewer

Best case is that ofBorg builds the package attributes mentioned in the commit message. Worst cases span from nothing being built, because an alias was given, or CI failures get ignored. 

Some attributes of course have a nixos test attached, and that is great. Most attributes don't, and it might not be feasible to make that happen. Starting some executable and interacting with it is still the most valuable test scenario for most packages.

> whether release notes have been added should be obvious from the diff
> so a checkbox for that serves primarily as a reminder to write release notes
> (still very valuable!)

I agree, that we could structure the PR template better, but note that it also serves as a checklist for the reviewing party, to make sure we don't forget basic things along the way.

> what tests did the contributor perform that aren't performed by the CI (manual checks, nixpkgs-review?)

I very often see PRs that come without any checkbox or description or motivation, and that bothers me, because it implies that basic due diligence wasn't done.

Anyway, I'm not opposed to trimming and/or reformatting the PR template to something that makes people want to interact with it more. But simply trimming, because arguing lots of things in it are implicit doesn't agree wit me.

**Comment by RaitoBezarius** [comment]
> Best case is that ofBorg builds the package attributes mentioned in the commit message. Worst cases span from nothing being built, because an alias was given, or CI failures get ignored.

> Some attributes of course have a nixos test attached, and that is great. Most attributes don't, and it might not be feasible to make that happen. Starting some executable and interacting with it is still the most valuable test scenario for most packages.

I would go further and say that if we want to go the CI route, it also means we introduce blocking status check whenever the attributes are not found / understood, etc. Which will make it massively more painful for everyone who doesn't understand the regex mechanism.

**Comment by evils** [comment]
another draft, with placeholder phrasing, /* comments */, and without links


## Changes
- [ ] are made with CONTRIBUTING.md in mind
- [ ] include changes worth documenting /* feels redundant this close to contrib.md, but should be here somewhere */
- [ ] work /* intentionally unspecific, maybe "work at runtime" or "don't break the functionality of the binaries"? */

### Description
- Move contributor instructions to CONTRIBUTING.md
- Use semantically valid H2 tags for level-2 headings.

## Ran (manually or in CI), as applicable:
- [ ] Package test(s) (`package-name.tests` and/or `checkPhase`)
- [ ] NixOS module test(s) (`nixosTests.affected-module` or `package-name.tests`)
- [ ] dependant packages (`nixpkgs-review`)

/*
`lib/tests` and `pkgs/test` seem esoteric enough that changes where those are relevant would have the contributor reading the docs and/or getting help from someone who knows about this stuff (contrib.md should still mention it)
In short, i never understood `lib/tests` and `pkgs/test` so don't feel like it's worth confusing most contributors with it
*/

### Ran on:
- Linux
  - [ ] x86-64
  - [ ] aarch64
- Darwin
  - [ ] x86-64
  - [ ] aarch64
  - [ ] with `sandbox = true`

**Comment by evils** [comment]
> it also serves as a checklist for the reviewing party

should the goal be for all applicable items to be checked at the end of the PR process?

so that it's up to the reviewer(s) (including CI and contributors) to do the things,
up to the contributor to keep the checklist up-to-date
(it would indeed be awesome if CI could do the updating),
and up to a merger to decide if enough has been done?

**Comment by l0b0** [comment]
Thank you all for the feedback! I'll try splitting this up and changing this into a more palatable change. As a first change, I've split out the test instructions into #238501.

Auto-ticking what's been done is outside the scope for this change, since that's a huge job compared to just improving the template a bit.

**Comment by l0b0** [comment]
Closing after splitting into more useful PRs.
